src/librbd/ManagedLock.cc

```diff
@@ -0,0 +1,661 @@
+  : m_ioctx(ioctx), m_cct(reinterpret_cast<CephContext *>(ioctx.cct())),
+    m_work_queue(work_queue),
+    m_oid(oid),
+    m_watcher(watcher),
+    m_lock(util::unique_lock_name("librbd::ManagedLock<I>::m_lock", this)),
+    m_state(STATE_UNLOCKED) {
+}
+
+template <typename I>
+ManagedLock<I>::~ManagedLock() {
+  Mutex::Locker locker(m_lock);
+  assert(m_state == STATE_SHUTDOWN || m_state == STATE_UNLOCKED);
+}
+
+template <typename I>
+bool ManagedLock<I>::is_lock_owner() const {
+  Mutex::Locker locker(m_lock);
+
+  bool lock_owner;
+
+  switch (m_state) {
+  case STATE_LOCKED:
+  case STATE_REACQUIRING:
+  case STATE_PRE_SHUTTING_DOWN:
+  case STATE_POST_ACQUIRING:
+  case STATE_PRE_RELEASING:
+    lock_owner = true;
+    break;
+  default:
+    lock_owner = false;
+    break;
+  }
+
+  ldout(m_cct, 20) << "=" << lock_owner << dendl;
+  return lock_owner;
+}
+
+template <typename I>
+void ManagedLock<I>::shut_down(Context *on_shut_down) {
+  ldout(m_cct, 10) << dendl;
+
+  Mutex::Locker locker(m_lock);
+  assert(!is_shutdown_locked());
+  execute_action(ACTION_SHUT_DOWN, on_shut_down);
+}
+
+template <typename I>
+void ManagedLock<I>::acquire_lock(Context *on_acquired) {
+  int r = 0;
+  {
+    Mutex::Locker locker(m_lock);
+    if (is_shutdown_locked()) {
+      r = -ESHUTDOWN;
+    } else if (m_state != STATE_LOCKED || !m_actions_contexts.empty()) {
+      ldout(m_cct, 10) << dendl;
+      execute_action(ACTION_ACQUIRE_LOCK, on_acquired);
+      return;
+    }
+  }
+
+  on_acquired->complete(r);
+}
+
+template <typename I>
+void ManagedLock<I>::try_acquire_lock(Context *on_acquired) {
+  int r = 0;
+  {
+    Mutex::Locker locker(m_lock);
+    if (is_shutdown_locked()) {
+      r = -ESHUTDOWN;
+    } else if (m_state != STATE_LOCKED || !m_actions_contexts.empty()) {
+      ldout(m_cct, 10) << dendl;
+      execute_action(ACTION_TRY_LOCK, on_acquired);
+      return;
+    }
+  }
+
+  on_acquired->complete(r);
+}
+
+template <typename I>
+void ManagedLock<I>::release_lock(Context *on_released) {
+  int r = 0;
+  {
+    Mutex::Locker locker(m_lock);
+    if (is_shutdown_locked()) {
+      r = -ESHUTDOWN;
+    } else if (m_state != STATE_UNLOCKED || !m_actions_contexts.empty()) {
+      ldout(m_cct, 10) << dendl;
+      execute_action(ACTION_RELEASE_LOCK, on_released);
+      return;
+    }
+  }
+
+  on_released->complete(r);
+}
+
+template <typename I>
+void ManagedLock<I>::reacquire_lock(Context *on_reacquired) {
+  {
+    Mutex::Locker locker(m_lock);
+
+    if (m_state == STATE_WAITING_FOR_REGISTER) {
+      // restart the acquire lock process now that watch is valid
+      ldout(m_cct, 10) << ": " << "woke up waiting acquire" << dendl;
+      Action active_action = get_active_action();
+      assert(active_action == ACTION_TRY_LOCK ||
+             active_action == ACTION_ACQUIRE_LOCK);
+      execute_next_action();
+    } else if (!is_shutdown_locked() &&
+        (m_state == STATE_LOCKED || m_state == STATE_ACQUIRING)) {
```

Comment:
I think you are missing some states here: STATE_POST_ACQUIRING, STATE_WAITING_FOR_LOCK